### PR TITLE
MDNS Refactoring and Multi-Node

### PR DESCRIPTION
**Pull request by Apollon77**
This PR is implementing summarized the following:
* Enhance MDNS Scanner and Broadcaster to support multiple queries and announcements for diffrerent Instances/ports
* Adjust low and high level API to use the new scanners and broadcasters and allow to run multiple nodes in one node.js instance
* Adds MultiNodeDevice example file for multiple device nodes in one process
* Enhance MDNS Bradcaster to allow to specify all announcement details and remove all hardcoded things (only TCP support is still hardcoded on false)
* It also contains several fixes on MDNS level ( IPV6 zero-block decoding was wring, empty name encoding was wrong ...)
* Adds a sleep method to the time class
* Enhance controller to discover devices rather then expecting static ip/port data. Found addresses and devices are tried one afte rthe other (also passcode only pairing is so possible where all instances found are tried with one passcode until first working was found.

### CODE REVIEW

File: packages/matter.js/src/mdns/MdnsBroadcaster.ts [hunk 1]
```diff
@@ -137,7 +140,11 @@ export class MdnsBroadcaster {
                     PtrRecord(MATTER_SERVICE_QNAME, deviceMatterQname),
                     PtrRecord(fabricQname, deviceMatterQname),
                     SrvRecord(deviceMatterQname, { priority: 0, weight: 0, port: announcedNetPort, target: hostname }),
-                    TxtRecord(deviceMatterQname, ["SII=5000", "SAI=300", "T=1"]),
+                    TxtRecord(deviceMatterQname, [
+                        `SII=${sleepIdleInterval ?? 5000}`,  /* Sleepy Idle Interval */
```
Comment (lauckhart) [inline packages/matter.js/src/mdns/MdnsBroadcaster.ts:144]: Constants for these defaults (5000 & 300) may be overkill but if you destructure operationalInstanceData above you could have `{ sleepIdleInterval = 5000, sleepActiveInterval = 300 } = operationalInstanceData` which would tighten up the code and I think do a good job of documenting the defaults on the API

**Comment by lauckhart** [comment]
Nice, big step forward in functionality!  Code is easily digestible too.  I added some comments, mostly small stuff except see my comment r.e. random port selection.  I'd let the kernel take care of that.

**Review by lauckhart — CHANGES_REQUESTED** [review]

File: packages/matter.js/src/net/fake/UdpChannelFake.ts [hunk 2]
```diff
@@ -22,12 +22,15 @@ export class UdpChannelFake implements UdpChannel {
...
+    private readonly listeningPort: number;
 
     constructor(
         private readonly localAddress: string,
         private readonly listeningAddress: string | undefined,
-        private readonly listeningPort: number,
-    ) { }
+        listeningPort?: number,
+    ) {
+        this.listeningPort = listeningPort ?? 1024 + Math.random() * 64511; // Random port 1024-65535
```
[the remark this answers is not in the document]
Comment (Apollon77) [reply packages/matter.js/src/net/fake/UdpChannelFake.ts:32]: That's just the Fake!!Network one which is not real network at all ;-))

File: packages/matter.js/src/MatterController.ts [hunk 3]
```diff
@@ -77,20 +77,28 @@ export class MatterController {
...
         private readonly certificateManager: RootCertificateManager,
         private readonly fabric: Fabric,
         private readonly storage: StorageContext,
+        public operationalServerAddress?: ServerAddress,
         commissioningOptions?: CommissioningData
     ) {
         this.controllerStorage = this.storage.createContext("MatterController");
 
+        // If controller has a stored operational server address, use it, irrelevant what was passed in the constructor
+        if (this.controllerStorage.has("operationalServerAddress")) {
```
[the remark this answers is not in the document]
Comment (Apollon77) [reply packages/matter.js/src/MatterController.ts:105]: That storage entry is updated  whenever we have a "successful" connection.

File: packages/matter.js/src/net/fake/UdpChannelFake.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (lauckhart) [reply packages/matter.js/src/net/fake/UdpChannelFake.ts:32]: Hah duh

**Review by lauckhart — APPROVED** [review]

File: packages/matter.js/src/mdns/MdnsScanner.ts [hunk 4]
```diff
@@ -27,68 +52,618 @@ export class MdnsScanner implements Scanner {
...
+            }
+        }
+
+        await this.multicastServer.send(queryMessage);
+    }
+
+    /**
+     * Set new DnsQuery records to the list of active queries and start sending them out. When entry already exists the
+     * query is overwritten and answers are always added.
+     *
```
[the remark this answers is not in the document]
Comment (Apollon77) [reply packages/matter.js/src/mdns/MdnsScanner.ts:143]: Will try to add

File: packages/matter-node.js/src/net/UdpChannelNode.ts [hunk 5]
```diff
@@ -22,6 +22,11 @@ function createDgramSocket(address: string | undefined, port: number | undefined
         };
         socket.on("error", handleBindError);
         socket.bind(port, address, () => {
+            const { address: localAddress, port: localPort } = socket.address();
+            logger.debug("Socket created and bound ", Logger.dict({
+                remoteAddress: `${address}:${port}`,
```
[the remark this answers is not in the document]
Comment (Apollon77) [reply packages/matter-node.js/src/net/UdpChannelNode.ts:27]: changed some places

File: packages/matter.js/src/MatterController.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (Apollon77) [reply packages/matter.js/src/MatterController.ts:105]: Added as own method now

File: packages/matter.js/src/MatterController.ts [hunk 6]
```diff
@@ -184,21 +196,36 @@ export class MatterController {
...
-    async pair(peerNodeId: NodeId, operationalIp: string, operationalPort: number) {
+    async pair(peerNodeId: NodeId, operationalServerAddress: ServerAddress) {
+        const { ip: operationalIp, port: operationalPort } = operationalServerAddress;
         // Do CASE pairing
-        const operationalInterface = isIPv6(operationalIp) ? this.netInterfaceIpv6 : this.netInterfaceIpv4;
+        const isIpv6Address = isIPv6(operationalIp);
+        const operationalInterface = isIpv6Address ? this.netInterfaceIpv6 : this.netInterfaceIpv4;
+
+        if (operationalInterface === undefined) {
+            throw new PairRetransmissionLimitReachedError(`IPv${isIpv6Address ? "6" : "4"} interface not initialized for port ${operationalPort}. Can not use this address for pairing.`);
```
[the remark this answers is not in the document]
Comment (Apollon77) [reply packages/matter.js/src/MatterController.ts:377]: added

File: packages/matter.js/src/CommissioningController.ts [hunk 7]
```diff
@@ -108,17 +109,31 @@ export class CommissioningController extends MatterNode {
...
         } else {
-            this.nodeId = await this.controllerInstance.commission(this.ip, this.port, this.discriminator, this.passcode);
+            let identifierData;
+            if (this.longDiscriminator !== undefined) {
+                identifierData = { longDiscriminator: this.longDiscriminator };
+            } else if (this.shortDiscriminator !== undefined) {
+                identifierData = { shortDiscriminator: this.shortDiscriminator };
+            } else {
+                if (this.passcode === undefined) {
+                    throw new Error("Passcode not set!");
```
[the remark this answers is not in the document]
Comment (Apollon77) [reply packages/matter.js/src/CommissioningController.ts:129]: rephrased - and no.

File: packages/matter.js/src/MatterServer.ts [hunk 8]
```diff
@@ -82,30 +83,46 @@ export class MatterServer {
...
-                    await node.advertise();
-                }
-            } else if (node instanceof CommissioningController) {
-                node.setMdnsScanner(await MdnsScanner.create());
-                if (!node.delayedPairing) {
-                    await node.connect();
-                }
+            this.prepareNode(node);
+
+            if (node instanceof CommissioningServer && !node.delayedAnnouncement) {
```
[the remark this answers is not in the document]
Comment (Apollon77) [reply packages/matter.js/src/MatterServer.ts:96]: reafactored

File: packages/matter.js/src/MatterServer.ts [hunk 9]
```diff
@@ -82,30 +83,46 @@ export class MatterServer {
...
             }
         }
     }
 
+    private prepareNode(node: MatterNode) {
+        if (this.mdnsBroadcaster === undefined || this.mdnsScanner === undefined) {
+            logger.debug("Mdns instances not yet created, delaying node preparation");
+            return;
+        }
+        if (node instanceof CommissioningServer) {
```
[the remark this answers is not in the document]
Comment (Apollon77) [reply packages/matter.js/src/MatterServer.ts:109]: refactored

[the remark this answers is not in the document]
Comment (Apollon77) [reply packages/matter.js/src/MatterServer.ts:109]: (decided to leave it two calls for now)

File: packages/matter.js/src/CommissioningController.ts [hunk 10]
```diff
@@ -126,7 +127,7 @@ export class CommissioningController extends MatterNode {
                 identifierData = { shortDiscriminator: this.shortDiscriminator };
             } else {
                 if (this.passcode === undefined) {
-                    throw new Error("Passcode not set!");
+                    throw new Error("To commission a new device a passcode needs to be specified i the constructor data!");
```
Comment (lauckhart) [inline packages/matter.js/src/CommissioningController.ts:130]: Missing "n" I think

**Review by lauckhart — APPROVED** [review]
Looks good

File: packages/matter.js/src/CommissioningController.ts (hunk 10 above)
Comment (Apollon77) [reply packages/matter.js/src/CommissioningController.ts:130]: fixed

**Comment by Apollon77** [comment]
Tests fixed

**Pull request merged**